kernel: thread monitor - Assert reusing TCB in use

ChidvilasYQC opened the pull request:
Add a check in thread monitor to assert if a TCB is reused while it is still present in the kernel thread list.
This prevents undefined behavior due to TID reuse before proper cleanup.

tests: Add test for thread tcb reuse

peter-mitsis commented on tests/kernel/threads/reuse_tid/src/main.c:
```diff
@@ -0,0 +1,106 @@
+	printf("Giving test sem again\n");
+	k_sem_give(&test_sem);
+
+	k_thread_join(tid1, K_FOREVER);
```
What purpose do these lines serve?  If the test behaves as expected, then we never reach them. If it fails, then the failure should be caught by the subsequent ztest_test_fail().

ChidvilasYQC commented on kernel/thread.c:
```diff
@@ -657,6 +657,11 @@ char *z_setup_new_thread(struct k_thread *new_thread,
+	/* Check if the thread is already in the list */
+	for (struct k_thread *t = _kernel.threads; t; t = t->next_thread) {
+		__ASSERT(t != new_thread, "thread already in list");
+	}
```
Thanks for pointing that this loop is not protected under the lock.
1. Updated code to use `z_thread_monitor_lock`
2. This search is moved to the beginning of the `k_thread_create` implementation.
3. Since `z_thread_monitor_lock` must be unlocked before calling `__ASSERT`, I added an if condition to check for TID reuse. Inside this if block, the spinlock is released and the assertion is triggered.
Because the compiler will not optimize out `k_spin_unlock` inside the if block when assertions are disabled, I placed this search under `CONFIG_ASSERT`.

ChidvilasYQC commented:
> @ChidvilasYQC - Instead of appending new commits that address the review comments, please squash them into the relevant originating commit(s).

Squashed all the commits and rebased main branch.

nashif reviewed (changes requested):
is there a real problem this change is solving? what is the bug and when does this happen and why? 
asserts should not be enable in production, so this does not solve any problem really..

andyross reviewed (commented):
One slightly pedantic note about the lock.  Reviewed mostly just the kernel code, not the test.

Generally this looks reasonable to me.  It's a legitimate (and easy) mistake to make, and this test probably provides value.  There's a performance worry about adding the O(N) search of the global thread list on every thread creation, but in practice there aren't that many threads in a running system (the size of the stacks needed keeps these things down into the "few dozen" or so for realistic apps).

It's also been discussed that we could solve this in a heavier/less-compatible way with thread struct state: maybe require that the thread struct have a magic "has been destroyed" cookie set on it somewhere.  That would require the use of a new API like `k_thread_struct_init()` or whatever though.  But we could also grandfather in a rule like "zero-filled structs are OK" and probably make it for for 99% of existing code.

ChidvilasYQC commented on kernel/thread.c:
```diff
@@ -586,6 +586,28 @@ char *z_setup_new_thread(struct k_thread *new_thread,
+	/* Check if the thread is already in the list */
+	for (struct k_thread *t = _kernel.threads; t; t = t->next_thread) {
+		if (t == new_thread) {
+			k_spin_unlock(&z_thread_monitor_lock, key);
```
got it, I have removed spin unlock.

ChidvilasYQC commented on tests/kernel/threads/thread_error_case/src/main.c:
```diff
@@ -25,14 +28,21 @@ enum {
 	THREAD_CREATE_NEWTHREAD_NULL,
 	THREAD_CREATE_STACK_NULL,
-	THREAD_CTEATE_STACK_SIZE_OVERFLOW
+	THREAD_CTEATE_STACK_SIZE_OVERFLOW,
```
typo fixed.

ChidvilasYQC commented:
@peter-mitsis, @nashif and @andyross can you please review

ChidvilasYQC commented:
Hi @peter-mitsis, shall I close this PR and create a new one to fix issues with PR checks?

peter-mitsis commented:
@ChidvilasYQC - Near as I can tell, the CI failures are unrelated to the changes. My recommendation at this point would be ...
1. If you know how to fix the cause of the failures, submit a separate PR to fix them.
2. When fixed (or if already fixed), rebase and re-push to re-run CI.

peter-mitsis reviewed (dismissed):
Note to any other reviewers: The CI failures are unrelated to this change set.

npitre commented on kernel/thread.c:
```diff
@@ -589,6 +589,27 @@ char *z_setup_new_thread(struct k_thread *new_thread,
+	new_thread->entry.parameter3 = p3;
+
+	new_thread->next_thread = _kernel.threads;
+	_kernel.threads = new_thread;
```
With this block moved to the top of `z_setup_new_thread()`, the new thread is linked into `_kernel.threads` **before** `z_waitq_init()`, `z_init_thread_base()`, `setup_thread_stack()`, `arch_new_thread()`, and the rest of the initialization run further down. Anything iterating the list on another CPU — `k_thread_foreach`, `k_thread_foreach_filter_by_cpu`, a debug shell, the thread analyzer — will now observe a thread with uninitialized `base` (no valid `thread_state`/`prio`/`cpu_mask`), NULL/stale `switch_handle`, etc. The previous code only published the thread after full init for exactly this reason.

Suggest keeping the insertion at the original (end-of-routine) location, and doing the check + insert atomically under the same lock there:

```c
#ifdef CONFIG_THREAD_MONITOR
    k_spinlock_key_t key = k_spin_lock(&z_thread_monitor_lock);

#ifdef CONFIG_ASSERT
    for (struct k_thread *t = _kernel.threads; t != NULL; t = t->next_thread) {
        __ASSERT(t != new_thread,
                 "thread %p is already in the running list", new_thread);
    }
#endif

    new_thread->entry.pEntry     = entry;
    new_thread->entry.parameter1 = p1;
    new_thread->entry.parameter2 = p2;
    new_thread->entry.parameter3 = p3;

    new_thread->next_thread = _kernel.threads;
    _kernel.threads = new_thread;
    k_spin_unlock(&z_thread_monitor_lock, key);
#endif
```

Same atomicity guarantee, but no partially-initialized thread is ever visible in the global list. If you specifically want to fail fast on reuse *before* clobbering the caller's TCB, do a separate check-only walk under the lock at the top and keep the insert at the bottom — but the assert is best-effort either way, since two CPUs racing on the same TCB will already be clobbering each other before either's check runs.

ChidvilasYQC replied:
@npitre, this code is protected by the `z_thread_monitor_lock`, hence other utilities like shell thread commands must use this lock before reading thread list.

Also, this suggestion from @peter-mitsis makes sense. We cannot let `new_thread` values to be overwritten.
>The search should be moved up to the start of the routine. Otherwise if there is reuse, then we run the risk of corrupting data before we find out that it was in use.

npitre replied:
The lock isn't the issue I'm raising. `z_thread_monitor_lock` serializes access to the list *pointers*, so a reader won't trip over a half-linked `next_thread` — agreed. But it doesn't order the initialization of `new_thread`'s own fields.

With both the check and the insert (`_kernel.threads = new_thread`) at the top, the thread is published into the global list before `z_init_thread_base()`, `arch_new_thread()`, and the rest run below. A `k_thread_foreach` / thread-analyzer / shell walk on another CPU that correctly takes the lock will find `new_thread` and then read `base.thread_state`, `base.prio`, `base.cpu_mask`, `switch_handle` — all still garbage at that point. The lock is held correctly on both sides; the data simply isn't initialized yet.

To get both properties — fail-fast on reuse *before* clobbering the caller's TCB, and no partially-initialized thread visible in the list — split it: a check-only walk at the top under the lock (no insert), and keep the insert at the bottom after the thread is fully built, under the lock again. The reuse `__ASSERT` fires from the top walk; publication still happens last.

npitre commented on kernel/thread.c:
```diff
@@ -589,6 +589,27 @@ char *z_setup_new_thread(struct k_thread *new_thread,
+	/* Check if the thread is already in the list */
+	for (struct k_thread *t = _kernel.threads; t; t = t->next_thread) {
+		if (t == new_thread) {
+			__ASSERT(0, "thread %p is already in the running list\n", new_thread);
```
The `if (...) __ASSERT(0, ...)` form is unusual; the idiom elsewhere in this file is to fold the condition into the assert itself:

```c
#ifdef CONFIG_ASSERT
    for (struct k_thread *t = _kernel.threads; t != NULL; t = t->next_thread) {
        __ASSERT(t != new_thread,
                 "thread %p is already in the running list", new_thread);
    }
#endif
```

Keep the `#ifdef CONFIG_ASSERT` (or `if (IS_ENABLED(CONFIG_ASSERT))`) wrapper — the compiler won't eliminate the list walk on its own, since it reads through `_kernel.threads` / `next_thread`.

Minor: the trailing `\n` in the message isn't the style used by other `__ASSERT` calls in this file.

ChidvilasYQC replied:
Right, I was having spin unlock before assert, so I had to use if condition check.

But after this suggestion from @andyross, I have removed the spin unlock. But kept the if condition. Since, there is only one assert statement now, I will remove if condition.
>This line will produce a double-unlock in the situation where the app has a custom assert handler that doesn't panic. And obviously, in the default case where assert does panic, you don't care about lock hygiene. Asserts and panics work just fine when you're holding a spinlock.
Remove, basically.

npitre commented on tests/kernel/threads/thread_error_case/src/main.c:
```diff
@@ -95,6 +105,24 @@ static void tThread_entry_negative(void *p1, void *p2, void *p3)
+			K_PRIO_PREEMPT(THREAD_TEST_PRIORITY + 1),
+			perm, K_NO_WAIT);
+		/* Give the thread time to start */
+		k_sleep(K_MSEC(10));
```
Is the `k_sleep(K_MSEC(10))` needed here? `k_thread_create()` adds the new thread to `_kernel.threads` synchronously (kernel/thread.c around the monitor-lock block), and the first thread's lower priority (`THREAD_TEST_PRIORITY + 1`) means the caller isn't preempted — so the second `k_thread_create` finds `reuse_tdata` on the list and asserts whether or not the first thread has ever been scheduled in. `k_sleep` doesn't remove a thread from the monitor list either; only `z_thread_monitor_exit` does, on abort.

If that's right, dropping the sleep and making `long_running_thread` just `k_sleep(K_FOREVER)` would remove a timing dependency from the test.

ChidvilasYQC replied:
updated with for ever sleep

npitre reviewed (changes requested):
Approach is reasonable, but the new monitor-list insertion at the top of `z_setup_new_thread()` publishes the thread to `_kernel.threads` before its `base`, stack and arch state are initialized — see inline comment. A couple of style/test nits too.

ChidvilasYQC committed:
kernel: thread monitor - Assert reusing TCB in use

Add a check in thread monitor to assert if a TCB is reused while it is
still present in the kernel thread list.
This prevents undefined behavior due to TID reuse before proper cleanup.

Signed-off-by: Chidvilas Yerramsetti <cyerrams@qti.qualcomm.com>

ChidvilasYQC commented:
@npitre, moved reuse tid check into a new function and it happens at the start of z_setup_new_thread

npitre commented on kernel/thread.c:
```diff
@@ -873,6 +873,26 @@ static inline void init_thread_usage(struct k_thread *thread)
+	/* Check if the thread is already in the list */
+	for (struct k_thread *t = _kernel.threads; t; t = t->next_thread) {
+		if (t == new_thread) {
+			k_spin_unlock(&z_thread_monitor_lock, key);
```
This unlock inside the `if`, together with the unconditional `k_spin_unlock` at the end of the function, double-unlocks `z_thread_monitor_lock` on reuse when the assert handler returns (a custom, non-panicking handler): unlock here, `__ASSERT` returns, the loop falls through to the trailing unlock. This is the block andyross asked to drop and you replied you would remove. Delete the `if (t == new_thread)` unlock and keep only the folded `__ASSERT` plus the single trailing unlock.

ChidvilasYQC replied:
Fixed both points from the reivew:
 - Double-unlock: removed the k_spin_unlock inside the if (t == new_thread) branch. check_if_tid_is_reused() now just sets a reused flag and breaks out of the loop, unlocks once unconditionally after the loop, then asserts on !reused. So the lock is released exactly once regardless of whether reuse is detected.
- K_MSEC(10) sleep: removed - the reuse check in z_setup_new_thread() runs synchronously on the second k_thread_create() call, so there's nothing to wait for.

npitre reviewed (changes requested):
The double-unlock andyross flagged is back (see inline): the `if (t == new_thread) { k_spin_unlock(...); }` inside the loop plus the new unconditional unlock at the end of the function release the lock twice when reuse is detected and the assert handler returns.

The publish-before-init concern is resolved now that the walk is check-only and the insert stays in z_thread_monitor_init(), and the assert-message idiom is fixed. Minor: the K_MSEC(10) sleep in THREAD_REUSE_ACTIVE_THREAD_ID is still there (only long_running_thread changed to K_FOREVER) and is unnecessary since the reuse check is synchronous.

npitre commented on tests/kernel/threads/thread_error_case/src/main.c:
```diff
@@ -288,6 +311,22 @@ ZTEST_USER(thread_error_case, test_thread_create_stack_overflow)
+ *
+ * @see k_thread_create()
+ */
+ZTEST_USER(thread_error_case, test_thread_reuse_active_thread_id)
```
Use `ZTEST` rather than `ZTEST_USER` here so the worker runs in supervisor mode and the create actually reaches `z_setup_new_thread()`. The `if (k_is_user_context())` block at lines 164-166 then becomes dead and can go, as can the three `reuse_*` arguments added to `k_thread_access_grant()` in `thread_grant_setup()`.

ChidvilasYQC replied:
Switched to `ZTEST` and dropped `reuse_tdata`/`reuse_stack1`/`reuse_stack2` from `k_thread_access_grant()` in `thread_grant_setup()`. Also removing the now-dead if `(k_is_user_context())` block.

npitre commented on kernel/thread.c:
```diff
@@ -873,6 +873,28 @@ static inline void init_thread_usage(struct k_thread *thread)
 #endif /* CONFIG_SCHED_THREAD_USAGE */
 }
 
+static inline void check_if_tid_is_reused(struct k_thread *new_thread)
```
Naming: the helpers around this one are `init_thread_*()` / `assert_*()` (see `assert_thread_coherence()`, `assert_switch_handle()`), so `assert_thread_not_reused()` would fit better, and `tid` is not a term used in this file. The two `#endif`s also want their `/* CONFIG_... */` comments like every other helper in this block, and the feature-disabled path needs an `ARG_UNUSED(new_thread)`.

ChidvilasYQC replied:
Renamed to `assert_thread_not_reused` to match the `assert_*` helpers, added the `/* CONFIG_ASSERT */` / `/* CONFIG_THREAD_MONITOR */` labels on the `#endif`s, and added `ARG_UNUSED(new_thread)` for the disabled path.

npitre reviewed (commented):
The added test never reaches the added assert. `thread_error_case` is filtered on `CONFIG_ARCH_HAS_USERSPACE` and sets `CONFIG_TEST_USERSPACE=y`, which force-enables `CONFIG_USERSPACE` (`TEST_ENABLE_USERSPACE` selects it), so `test_thread_reuse_active_thread_id` always runs in user mode. The second `k_thread_create()` is then rejected by `K_SYSCALL_OBJ_NEVER_INIT()` in `z_vrfy_k_thread_create()`, because the first create already ran `k_object_init()` on `reuse_tdata`, and `z_setup_new_thread()` is never entered. Running the suite on qemu_x86 at 87420c9:

```
START - test_thread_reuse_active_thread_id
E: 0x137000 k_thread in use
E: syscall z_vrfy_k_thread_create failed check: access denied
```

That is the same verifier check `test_thread_create_uninit` already covers, so the new kernel code ends up with no coverage in any configuration CI builds. Making the case supervisor-mode does reach it:

```
ASSERTION FAIL [!reused] @ ZEPHYR_BASE/kernel/thread.c:893
	thread 0x137000 is already in the running list
```

and the suite still passes, so `ztest_set_fault_valid(true)` already handles the assert panic and no assert hook is needed.

The double-unlock and the `K_MSEC(10)` sleep are both resolved.

npitre reviewed (approved):
Needs a rebase before it can merge

ChidvilasYQC commented:
@npitre, I have rebased main, please review and approve.
